// mppx/dist/Errors.js
/**
 * Base class for all payment-related errors.
 */
export class PaymentError extends Error {
    constructor() {
        super(...arguments);
        /** HTTP status code. */
        this.status = 402;
    }
    /** Converts the error to RFC 9457 Problem Details format. */
    toProblemDetails(challengeId) {
        return {
            type: this.type,
            title: this.title,
            status: this.status,
            detail: this.message,
            ...(challengeId && { challengeId }),
        };
    }
}
/**
 * Credential is malformed (invalid base64url, bad JSON structure).
 */
export class MalformedCredentialError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Credential is malformed: ${reason}.` : 'Credential is malformed.');
        this.name = 'MalformedCredentialError';
        this.title = 'Malformed Credential';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/malformed-credential';
    }
}
/**
 * Challenge ID is unknown, expired, or already used.
 */
export class InvalidChallengeError extends PaymentError {
    constructor(options = {}) {
        const { id, reason } = options;
        const idPart = id ? ` "${id}"` : '';
        const reasonPart = reason ? `: ${reason}` : '';
        super(`Challenge${idPart} is invalid${reasonPart}.`);
        this.name = 'InvalidChallengeError';
        this.title = 'Invalid Challenge';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/invalid-challenge';
    }
}
/**
 * Payment proof is invalid or verification failed.
 */
export class VerificationFailedError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Payment verification failed: ${reason}.` : 'Payment verification failed.');
        this.name = 'VerificationFailedError';
        this.title = 'Verification Failed';
        this.type = 'https://paymentauth.org/problems/verification-failed';
    }
}
/**
 * Payment requires additional action (e.g., 3DS authentication).
 */
export class PaymentActionRequiredError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Payment requires action: ${reason}.` : 'Payment requires action.');
        this.name = 'PaymentActionRequiredError';
        this.title = 'Payment Action Required';
        this.type = 'https://paymentauth.org/problems/payment-action-required';
    }
}
/**
 * Payment has expired.
 */
export class PaymentExpiredError extends PaymentError {
    constructor(options = {}) {
        const { expires } = options;
        super(expires ? `Payment expired at ${expires}.` : 'Payment has expired.');
        this.name = 'PaymentExpiredError';
        this.title = 'Payment Expired';
        this.type = 'https://paymentauth.org/problems/payment-expired';
    }
}
/**
 * No credential was provided but payment is required.
 */
export class PaymentRequiredError extends PaymentError {
    constructor(options = {}) {
        const { description } = options;
        const parts = ['Payment is required'];
        if (description)
            parts.push(`(${description})`);
        super(`${parts.join(' ')}.`);
        this.name = 'PaymentRequiredError';
        this.title = 'Payment Required';
        this.type = 'https://paymentauth.org/problems/payment-required';
    }
}
/**
 * Credential payload does not match the expected schema.
 */
export class InvalidPayloadError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Credential payload is invalid: ${reason}.` : 'Credential payload is invalid.');
        this.name = 'InvalidPayloadError';
        this.title = 'Invalid Payload';
        this.type = 'https://paymentauth.org/problems/invalid-payload';
    }
}
/**
 * Request is malformed or contains invalid parameters.
 */
export class BadRequestError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Bad request: ${reason}.` : 'Bad request.');
        this.name = 'BadRequestError';
        this.title = 'Bad Request';
        this.status = 400;
        this.type = 'https://paymentauth.org/problems/bad-request';
    }
}
/**
 * Payment amount is insufficient (too low).
 */
export class PaymentInsufficientError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Payment insufficient: ${reason}.` : 'Payment amount is insufficient.');
        this.name = 'PaymentInsufficientError';
        this.title = 'Payment Insufficient';
        this.type = 'https://paymentauth.org/problems/payment-insufficient';
    }
}
/**
 * Payment method is not supported by the server.
 */
export class PaymentMethodUnsupportedError extends PaymentError {
    constructor(options = {}) {
        const { method } = options;
        super(method ? `Payment method "${method}" is not supported.` : 'Payment method is not supported.');
        this.name = 'PaymentMethodUnsupportedError';
        this.title = 'Method Unsupported';
        this.status = 400;
        this.type = 'https://paymentauth.org/problems/method-unsupported';
    }
}
/**
 * Insufficient balance in the payment channel.
 */
export class InsufficientBalanceError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Insufficient balance: ${reason}.` : 'Insufficient balance.');
        this.name = 'InsufficientBalanceError';
        this.title = 'Insufficient Balance';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/session/insufficient-balance';
    }
}
/**
 * Voucher or close request signature is invalid.
 */
export class InvalidSignatureError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Invalid signature: ${reason}.` : 'Invalid signature.');
        this.name = 'InvalidSignatureError';
        this.title = 'Invalid Signature';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/session/invalid-signature';
    }
}
/**
 * Recovered signer is not authorized for this channel.
 */
export class SignerMismatchError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Signer mismatch: ${reason}.` : 'Signer is not authorized for this channel.');
        this.name = 'SignerMismatchError';
        this.title = 'Signer Mismatch';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/session/signer-mismatch';
    }
}
/**
 * Voucher cumulative amount exceeds the channel deposit.
 */
export class AmountExceedsDepositError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Amount exceeds deposit: ${reason}.` : 'Voucher amount exceeds channel deposit.');
        this.name = 'AmountExceedsDepositError';
        this.title = 'Amount Exceeds Deposit';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/session/amount-exceeds-deposit';
    }
}
/**
 * Voucher amount increase is below the minimum delta.
 */
export class DeltaTooSmallError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Delta too small: ${reason}.` : 'Amount increase below minimum voucher delta.');
        this.name = 'DeltaTooSmallError';
        this.title = 'Delta Too Small';
        this.status = 402;
        this.type = 'https://paymentauth.org/problems/session/delta-too-small';
    }
}
/**
 * No channel with this ID exists.
 */
export class ChannelNotFoundError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Channel not found: ${reason}.` : 'No channel with this ID exists.');
        this.name = 'ChannelNotFoundError';
        this.title = 'Channel Not Found';
        this.status = 410;
        this.type = 'https://paymentauth.org/problems/session/channel-not-found';
    }
}
/**
 * Channel is closed or finalized.
 */
export class ChannelClosedError extends PaymentError {
    constructor(options = {}) {
        const { reason } = options;
        super(reason ? `Channel closed: ${reason}.` : 'Channel is closed.');
        this.name = 'ChannelClosedError';
        this.title = 'Channel Closed';
        this.status = 410;
        this.type = 'https://paymentauth.org/problems/session/channel-finalized';
    }
}
