// mppx/dist/Store.js
/**
 * Async key-value store interface.
 *
 * Modeled after Cloudflare KV's API (`get`/`put`/`delete`).
 * Implementations handle serialization internally.
 */
import { Json } from 'ox';
/** Creates a {@link Store} from an existing implementation. */
export function from(store) {
    return store;
}
/** Wraps a Cloudflare KV namespace. */
export function cloudflare(kv) {
    return from({
        async get(key) {
            const raw = await kv.get(key);
            if (raw == null)
                return null;
            return Json.parse(raw);
        },
        async put(key, value) {
            await kv.put(key, Json.stringify(value));
        },
        async delete(key) {
            await kv.delete(key);
        },
    });
}
/** In-memory store backed by a `Map`. JSON-roundtrips values to match production behavior. */
export function memory() {
    const store = new Map();
    return from({
        async get(key) {
            const raw = store.get(key);
            if (raw === undefined)
                return null;
            return Json.parse(raw);
        },
        async put(key, value) {
            store.set(key, Json.stringify(value));
        },
        async delete(key) {
            store.delete(key);
        },
    });
}
/** Wraps an Upstash Redis instance (e.g. Vercel KV). */
export function upstash(redis) {
    return from({
        async get(key) {
            return (await redis.get(key));
        },
        async put(key, value) {
            await redis.set(key, value);
        },
        async delete(key) {
            await redis.del(key);
        },
    });
}
