// mppx/dist/proxy/Service.js
import { Value } from 'ox';
/**
 * Creates a service definition.
 *
 * @example
 * ```ts
 * Service.from('my-api', {
 *   baseUrl: 'https://api.example.com',
 *   bearer: 'sk-...',
 *   routes: {
 *     'POST /v1/generate': mppx.charge({ amount: '0.01' }),
 *     'GET /v1/status': true,
 *   },
 * })
 * ```
 */
export function from(id, config) {
    const rewriteFromConfig = resolveRewriteRequest(config);
    return {
        baseUrl: config.baseUrl,
        description: config.description,
        id,
        docsLlmsUrl: resolveLlmsUrl(config.docsLlmsUrl),
        routes: config.routes,
        title: config.title,
        rewriteRequest: config.rewriteRequest
            ? rewriteFromConfig
                ? async (req, ctx) => {
                    req = await rewriteFromConfig(req, ctx);
                    return config.rewriteRequest(req, ctx);
                }
                : config.rewriteRequest
            : rewriteFromConfig,
    };
}
export { from as custom };
function resolveRewriteRequest(config) {
    if (config.mutate) {
        const mutate = config.mutate;
        return (req, ctx) => {
            const options = ctx;
            const m = options.mutate ?? mutate;
            return m(req);
        };
    }
    if (config.bearer) {
        const bearer = config.bearer;
        return (req, ctx) => {
            const options = ctx;
            req.headers.set('Authorization', `Bearer ${options.bearer ?? bearer}`);
            return req;
        };
    }
    if (config.headers) {
        const headers = config.headers;
        return (req, ctx) => {
            const options = ctx;
            const h = options.headers ?? headers;
            for (const [name, value] of Object.entries(h))
                req.headers.set(name, value);
            return req;
        };
    }
    return undefined;
}
/** Serializes a service for discovery responses. */
export function serialize(s) {
    return {
        description: s.description,
        id: s.id,
        docsLlmsUrl: s.docsLlmsUrl?.({}),
        routes: Object.entries(s.routes).map(([pattern, endpoint]) => {
            const tokens = pattern.trim().split(/\s+/);
            const hasMethod = tokens.length >= 2;
            const path = hasMethod ? tokens.slice(1).join(' ') : tokens[0];
            return {
                docsLlmsUrl: s.docsLlmsUrl?.({ route: pattern }),
                method: hasMethod ? tokens[0] : undefined,
                path: `/${s.id}${path}`,
                pattern: hasMethod ? `${tokens[0]} /${s.id}${path}` : `/${s.id}${path}`,
                payment: endpoint ? resolvePayment(endpoint) : null,
            };
        }),
        title: s.title,
    };
}
/** Renders an llms.txt markdown string for a list of services. */
export function toLlmsTxt(services, options) {
    const lines = [
        `# ${options?.title ?? 'API Proxy'}`,
        '',
        `> ${options?.description ?? 'Paid API proxy powered by [Machine Payments Protocol](https://mpp.tempo.xyz).'}`,
        '',
    ];
    if (services.length === 0)
        return lines.join('\n');
    lines.push('## Services', '');
    for (const s of services) {
        const label = s.title ?? s.id;
        const desc = s.description ? `: ${s.description}` : '';
        lines.push(`- [${label}](/discover/${s.id}.md)${desc}`);
    }
    lines.push('', '[See all service definitions](/discover/all.md)');
    return lines.join('\n');
}
/** Renders a full markdown listing of all services with their routes. */
export function toServicesMarkdown(services) {
    const lines = ['# Services', ''];
    if (services.length === 0)
        return lines.join('\n');
    for (const s of services) {
        lines.push(`## [${s.title ?? s.id}](/discover/${s.id}.md)`, '');
        if (s.description)
            lines.push(s.description, '');
        pushRoutes(lines, s);
    }
    return lines.join('\n');
}
/** Renders a markdown string for a single service. */
export function toMarkdown(s) {
    const docsLlmsUrl = s.docsLlmsUrl?.({});
    const lines = [`# ${s.title ?? s.id}`, ''];
    if (docsLlmsUrl)
        lines.push(`> Documentation: ${docsLlmsUrl}`, '');
    if (s.description)
        lines.push(s.description, '');
    pushRoutes(lines, s, '##');
    return lines.join('\n');
}
function pushRoutes(lines, s, heading = '###') {
    lines.push(`${heading} Routes`, '');
    const serialized = serialize(s);
    for (const route of serialized.routes) {
        const p = route.payment;
        const desc = p?.description ? `: ${p.description}` : '';
        lines.push(`- \`${route.pattern}\`${desc}`);
        if (!p) {
            lines.push('  - Type: free');
        }
        else {
            lines.push(`  - Type: ${p.intent}`);
            if (p.amount) {
                const perUnit = p.unitType ? `/${p.unitType}` : '';
                if (p.decimals !== undefined) {
                    const price = Number(p.amount) / 10 ** Number(p.decimals);
                    lines.push(`  - Price: ${price}${perUnit} (${p.amount} units, ${p.decimals} decimals)`);
                }
                else {
                    lines.push(`  - Units: ${p.amount}${perUnit}`);
                }
            }
            if (p.currency)
                lines.push(`  - Currency: ${p.currency}`);
        }
        if (route.docsLlmsUrl)
            lines.push(`  - Docs: ${route.docsLlmsUrl}`);
        lines.push('');
    }
}
/** Extracts per-endpoint options from an endpoint definition. */
export function getOptions(endpoint) {
    if (typeof endpoint === 'object' && endpoint !== null && 'options' in endpoint)
        return endpoint.options;
    return undefined;
}
function resolvePayment(endpoint) {
    if (endpoint === true)
        return null;
    const handler = typeof endpoint === 'function' ? endpoint : endpoint.pay;
    if (!('_internal' in handler))
        return {};
    const { name, intent, defaults, schema, _canonicalRequest, ...rest } = handler._internal;
    const amount = (() => {
        if (typeof rest.amount === 'string' && typeof rest.decimals === 'number')
            return String(Value.from(rest.amount, rest.decimals));
        return rest.amount;
    })();
    return { intent, method: name, ...rest, ...(amount !== undefined && { amount }) };
}
function resolveLlmsUrl(input) {
    if (!input)
        return undefined;
    if (typeof input === 'function')
        return input;
    return ({ route }) => (route ? undefined : input);
}
