// mppx/dist/tempo/client/Charge.js
import { prepareTransactionRequest, sendCallsSync, signTransaction } from 'viem/actions';
import { tempo as tempo_chain } from 'viem/chains';
import { Actions } from 'viem/tempo';
import * as Credential from '../../Credential.js';
import * as Method from '../../Method.js';
import * as Account from '../../viem/Account.js';
import * as Client from '../../viem/Client.js';
import * as z from '../../zod.js';
import * as Attribution from '../Attribution.js';
import * as AutoSwap from '../internal/auto-swap.js';
import * as defaults from '../internal/defaults.js';
import * as Methods from '../Methods.js';
/**
 * Creates a Tempo charge method intent for usage on the client.
 *
 * @example
 * ```ts
 * import { tempo } from 'mppx/client'
 * import { privateKeyToAccount } from 'viem/accounts'
 *
 * const charge = tempo.charge({
 *   account: privateKeyToAccount('0x...'),
 * })
 * ```
 */
export function charge(parameters = {}) {
    const { clientId } = parameters;
    const getClient = Client.getResolver({
        chain: tempo_chain,
        getClient: parameters.getClient,
        rpcUrl: defaults.rpcUrl,
    });
    const getAccount = Account.getResolver({ account: parameters.account });
    return Method.toClient(Methods.charge, {
        context: z.object({
            account: z.optional(z.custom()),
            autoSwap: z.optional(z.custom()),
            mode: z.optional(z.enum(['push', 'pull'])),
        }),
        async createCredential({ challenge, context }) {
            const chainId = challenge.request.methodDetails?.chainId;
            const client = await getClient({ chainId });
            const account = getAccount(client, context);
            const mode = context?.mode ?? parameters.mode ?? (account.type === 'json-rpc' ? 'push' : 'pull');
            const { request } = challenge;
            const { amount, methodDetails } = request;
            const currency = request.currency;
            const recipient = request.recipient;
            const memo = methodDetails?.memo
                ? methodDetails.memo
                : Attribution.encode({ serverId: challenge.realm, clientId });
            const transferCall = Actions.token.transfer.call({
                amount: BigInt(amount),
                memo,
                to: recipient,
                token: currency,
            });
            const autoSwap = AutoSwap.resolve(context?.autoSwap ?? parameters.autoSwap, AutoSwap.defaultCurrencies);
            const swapCalls = autoSwap
                ? await AutoSwap.findCalls(client, {
                    account: account.address,
                    amountOut: BigInt(amount),
                    tokenOut: currency,
                    tokenIn: autoSwap.tokenIn,
                    slippage: autoSwap.slippage,
                })
                : undefined;
            const calls = [...(swapCalls ?? []), transferCall];
            if (mode === 'push') {
                const { receipts } = await sendCallsSync(client, {
                    account,
                    calls: calls,
                    experimental_fallback: true,
                });
                const hash = receipts?.[0]?.transactionHash;
                if (!hash)
                    throw new Error('No transaction receipt returned.');
                return Credential.serialize({
                    challenge,
                    payload: { hash, type: 'hash' },
                    source: `did:pkh:eip155:${chainId}:${account.address}`,
                });
            }
            const prepared = await prepareTransactionRequest(client, {
                account,
                calls,
                ...(methodDetails?.feePayer && { feePayer: true }),
                nonceKey: 'expiring',
            });
            // FIXME: figure out gas estimation issue for fee payer tx
            prepared.gas = prepared.gas + 5000n;
            const signature = await signTransaction(client, prepared);
            return Credential.serialize({
                challenge,
                payload: { signature, type: 'transaction' },
                source: `did:pkh:eip155:${chainId}:${account.address}`,
            });
        },
    });
}
