// mppx/dist/tempo/client/SessionManager.js
import * as Fetch from '../../client/internal/Fetch.js';
import { deserializeSessionReceipt } from '../session/Receipt.js';
import { parseEvent } from '../session/Sse.js';
import { session as sessionPlugin } from './Session.js';
/**
 * Creates a session manager that handles the full client payment lifecycle:
 * channel open, incremental vouchers, SSE streaming, and channel close.
 *
 * Internally delegates to the `session()` method for all
 * channel state management and credential creation, and to `Fetch.from`
 * for the 402 challenge/retry flow.
 *
 * ## Session resumption
 *
 * All channel state is held **in memory**. If the client process restarts,
 * the session is lost and a new on-chain channel will be opened on the next
 * request — the previous channel's deposit is orphaned until manually closed.
 *
 * When the server includes a `channelId` in the 402 challenge `methodDetails`,
 * the client will attempt to recover the channel by reading its on-chain state
 * via `getOnChainChannel()`. If the channel has a positive deposit and is not
 * finalized, it resumes from the on-chain settled amount.
 */
export function sessionManager(parameters) {
    const fetchFn = parameters.fetch ?? globalThis.fetch;
    let channel = null;
    let lastChallenge = null;
    let lastUrl = null;
    let spent = 0n;
    const method = sessionPlugin({
        account: parameters.account,
        authorizedSigner: parameters.authorizedSigner,
        getClient: parameters.client ? () => parameters.client : parameters.getClient,
        escrowContract: parameters.escrowContract,
        decimals: parameters.decimals,
        maxDeposit: parameters.maxDeposit,
        onChannelUpdate(entry) {
            if (entry.channelId !== channel?.channelId)
                spent = 0n;
            channel = entry;
        },
    });
    const wrappedFetch = Fetch.from({
        fetch: fetchFn,
        methods: [method],
        onChallenge: async (challenge, _helpers) => {
            lastChallenge = challenge;
            return undefined;
        },
    });
    function updateSpentFromReceipt(receipt) {
        if (!receipt || receipt.channelId !== channel?.channelId)
            return;
        const next = BigInt(receipt.spent);
        spent = spent > next ? spent : next;
    }
    function toPaymentResponse(response) {
        const receiptHeader = response.headers.get('Payment-Receipt');
        const receipt = receiptHeader ? deserializeSessionReceipt(receiptHeader) : null;
        updateSpentFromReceipt(receipt);
        return Object.assign(response, {
            receipt,
            challenge: lastChallenge,
            channelId: channel?.channelId ?? null,
            cumulative: channel?.cumulativeAmount ?? 0n,
        });
    }
    async function doFetch(input, init) {
        lastUrl = input;
        const response = await wrappedFetch(input, init);
        return toPaymentResponse(response);
    }
    const self = {
        get channelId() {
            return channel?.channelId;
        },
        get cumulative() {
            return channel?.cumulativeAmount ?? 0n;
        },
        get opened() {
            return channel?.opened ?? false;
        },
        async open(options) {
            if (channel?.opened)
                return;
            if (!lastChallenge) {
                throw new Error('No challenge available. Make a request first to receive a 402 challenge, or pass a challenge via .fetch()/.sse().');
            }
            const deposit = options?.deposit;
            const credential = await method.createCredential({
                challenge: lastChallenge,
                context: {
                    ...(deposit !== undefined && { depositRaw: deposit.toString() }),
                },
            });
            if (!lastUrl)
                throw new Error('No URL available — call fetch() or sse() before open().');
            const response = await fetchFn(lastUrl, {
                method: 'POST',
                headers: { Authorization: credential },
            });
            if (!response.ok) {
                const body = await response.text().catch(() => '');
                const wwwAuth = response.headers.get('WWW-Authenticate') ?? '';
                throw new Error(`Open request failed with status ${response.status}${body ? `: ${body}` : ''}${wwwAuth ? ` [WWW-Authenticate: ${wwwAuth}]` : ''}`);
            }
        },
        fetch: doFetch,
        async sse(input, init) {
            const { onReceipt, signal, ...fetchInit } = init ?? {};
            const sseInit = {
                ...fetchInit,
                headers: {
                    ...fetchInit.headers,
                    Accept: 'text/event-stream',
                },
                ...(signal ? { signal } : {}),
            };
            const response = await doFetch(input, sseInit);
            // Snapshot the challenge at SSE open time so concurrent
            // calls don't overwrite it.
            const sseChallenge = lastChallenge;
            if (!response.body)
                throw new Error('Response has no body.');
            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            async function* iterate() {
                let buffer = '';
                try {
                    while (true) {
                        if (signal?.aborted)
                            break;
                        const { done, value } = await reader.read();
                        if (done)
                            break;
                        buffer += decoder.decode(value, { stream: true });
                        const parts = buffer.split('\n\n');
                        buffer = parts.pop();
                        for (const part of parts) {
                            if (!part.trim())
                                continue;
                            const event = parseEvent(part);
                            if (!event)
                                continue;
                            switch (event.type) {
                                case 'message':
                                    yield event.data;
                                    break;
                                case 'payment-need-voucher': {
                                    if (!channel || !sseChallenge)
                                        break;
                                    const required = BigInt(event.data.requiredCumulative);
                                    channel.cumulativeAmount =
                                        channel.cumulativeAmount > required ? channel.cumulativeAmount : required;
                                    const credential = await method.createCredential({
                                        challenge: sseChallenge,
                                        context: {
                                            action: 'voucher',
                                            channelId: channel.channelId,
                                            cumulativeAmountRaw: channel.cumulativeAmount.toString(),
                                        },
                                    });
                                    const voucherResponse = await fetchFn(input, {
                                        method: 'POST',
                                        headers: { Authorization: credential },
                                    });
                                    if (!voucherResponse.ok) {
                                        throw new Error(`Voucher POST failed with status ${voucherResponse.status}`);
                                    }
                                    break;
                                }
                                case 'payment-receipt':
                                    updateSpentFromReceipt(event.data);
                                    onReceipt?.(event.data);
                                    break;
                            }
                        }
                    }
                }
                finally {
                    reader.releaseLock();
                }
            }
            return iterate();
        },
        async close() {
            if (!channel?.opened || !lastChallenge)
                return undefined;
            const credential = await method.createCredential({
                challenge: lastChallenge,
                context: {
                    action: 'close',
                    channelId: channel.channelId,
                    cumulativeAmountRaw: spent.toString(),
                },
            });
            let receipt;
            if (lastUrl) {
                const response = await fetchFn(lastUrl, {
                    method: 'POST',
                    headers: { Authorization: credential },
                });
                const receiptHeader = response.headers.get('Payment-Receipt');
                if (receiptHeader)
                    receipt = deserializeSessionReceipt(receiptHeader);
            }
            return receipt;
        },
    };
    return self;
}
