// mppx/dist/server/Mppx.js
import * as Challenge from '../Challenge.js';
import * as Credential from '../Credential.js';
import * as Errors from '../Errors.js';
import * as Expires from '../Expires.js';
import * as Env from '../internal/env.js';
import * as PaymentRequest from '../PaymentRequest.js';
import * as NodeListener from './NodeListener.js';
import * as Request from './Request.js';
import * as Transport from './Transport.js';
/**
 * Creates a server-side payment handler from methods.
 *
 * It is highly recommended to set a `secretKey` to bind challenges to their contents,
 * and allow the server to verify that incoming credentials match challenges it issued.
 *
 * @example
 * ```ts
 * import { Mppx, tempo } from 'mppx/server'
 *
 * const payment = Mppx.create({
 *   methods: [tempo()],
 *   secretKey: process.env.PAYMENT_SECRET_KEY,
 * })
 * ```
 */
export function create(config) {
    const { realm = Env.get('realm') ?? 'MPP Payment', secretKey = Env.get('secretKey'), transport = Transport.http(), } = config;
    if (!secretKey) {
        throw new Error('Missing secret key. Set the MPP_SECRET_KEY environment variable or pass `secretKey` to Mppx.create().');
    }
    const methods = config.methods.flat();
    const handlers = {};
    const intentCount = {};
    for (const mi of methods) {
        intentCount[mi.intent] = (intentCount[mi.intent] ?? 0) + 1;
        handlers[`${mi.name}/${mi.intent}`] = createMethodFn({
            defaults: mi.defaults,
            method: mi,
            realm,
            request: mi.request,
            respond: mi.respond,
            secretKey,
            transport: (mi.transport ?? transport),
            verify: mi.verify,
        });
    }
    // Also set shorthand intent key when there's no collision
    for (const mi of methods) {
        if (intentCount[mi.intent] === 1)
            handlers[mi.intent] = handlers[`${mi.name}/${mi.intent}`];
    }
    // Build nested handlers: mppx.tempo.charge(...)
    for (const mi of methods) {
        if (!handlers[mi.name])
            handlers[mi.name] = {};
        const fn = handlers[`${mi.name}/${mi.intent}`];
        fn._method = mi;
        handlers[mi.name][mi.intent] = fn;
    }
    function composeFn(...entries) {
        if (transport.name !== 'http')
            throw new Error('compose() only supports HTTP transport');
        if (entries.length === 0)
            throw new Error('compose() requires at least one entry');
        const configured = entries.map(([methodOrKey, options]) => {
            const key = typeof methodOrKey === 'string'
                ? methodOrKey
                : typeof methodOrKey === 'function' && '_method' in methodOrKey
                    ? `${methodOrKey._method.name}/${methodOrKey._method.intent}`
                    : `${methodOrKey.name}/${methodOrKey.intent}`;
            const handlerFn = handlers[key];
            if (!handlerFn)
                throw new Error(`No handler for "${key}". Is this method in your methods array?`);
            return handlerFn(options);
        });
        return compose(...configured);
    }
    return {
        methods,
        compose: composeFn,
        realm: realm,
        transport,
        ...handlers,
    };
}
// biome-ignore lint/correctness/noUnusedVariables: _
function createMethodFn(parameters) {
    const { defaults, method, realm, respond, secretKey, transport, verify } = parameters;
    return (options) => {
        const { description, meta, ...rest } = options;
        const merged = { ...defaults, ...rest };
        return Object.assign(async (input) => {
            const expires = 'expires' in options ? options.expires : Expires.minutes(5);
            // Extract credential once — getCredential may have side effects (e.g. SSE transports).
            const [credential, credentialError] = (() => {
                try {
                    return [
                        transport.getCredential(input),
                        undefined,
                    ];
                }
                catch (e) {
                    return [null, e];
                }
            })();
            // Transform request if method provides a `request` function.
            const request = (parameters.request
                ? await parameters.request({ credential, request: merged })
                : merged);
            // Recompute challenge from options. The HMAC-bound ID means we don't need to
            // store challenges server-side—if the client echoes back a credential with
            // a matching ID, we know it was issued by us with these exact parameters.
            const challenge = Challenge.fromMethod(method, {
                description,
                expires,
                meta,
                realm,
                request,
                secretKey,
            });
            // Credential was provided but malformed
            if (credentialError) {
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error: new Errors.MalformedCredentialError({ reason: credentialError.message }),
                });
                return { challenge: response, status: 402 };
            }
            // No credential provided—issue challenge
            if (!credential) {
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error: new Errors.PaymentRequiredError({ description }),
                });
                return { challenge: response, status: 402 };
            }
            // Verify the echoed challenge was issued by us by recomputing its HMAC.
            // This is stateless—no database lookup needed.
            if (!Challenge.verify(credential.challenge, { secretKey })) {
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error: new Errors.InvalidChallengeError({
                        id: credential.challenge.id,
                        reason: 'challenge was not issued by this server',
                    }),
                });
                return { challenge: response, status: 402 };
            }
            // Verify the credential's challenge matches this route's configured
            // method, intent, realm, and request. Prevents cross-route scope
            // confusion where a credential issued for a cheap route (or different
            // method/intent) is presented at an expensive route.
            // Note: we compare specific payment parameters rather than the full
            // request because the `request` hook may produce credential-dependent
            // output (e.g. `feePayer` differs between 402 and credential calls).
            {
                for (const field of ['method', 'intent', 'realm']) {
                    if (credential.challenge[field] !== challenge[field]) {
                        const response = await transport.respondChallenge({
                            challenge,
                            input,
                            error: new Errors.InvalidChallengeError({
                                id: credential.challenge.id,
                                reason: `credential ${field} does not match this route's requirements`,
                            }),
                        });
                        return { challenge: response, status: 402 };
                    }
                }
                const routeReq = challenge.request;
                const echoedReq = credential.challenge.request;
                const routeDetails = (routeReq.methodDetails ?? {});
                const echoedDetails = (echoedReq.methodDetails ?? {});
                for (const field of ['amount', 'currency', 'recipient']) {
                    const routeVal = routeReq[field] ?? routeDetails[field];
                    if (routeVal !== undefined &&
                        String(routeVal) !== String(echoedReq[field] ?? echoedDetails[field])) {
                        const response = await transport.respondChallenge({
                            challenge,
                            input,
                            error: new Errors.InvalidChallengeError({
                                id: credential.challenge.id,
                                reason: `credential ${field} does not match this route's requirements`,
                            }),
                        });
                        return { challenge: response, status: 402 };
                    }
                }
            }
            // Reject expired credentials
            if (credential.challenge.expires && new Date(credential.challenge.expires) < new Date()) {
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error: new Errors.PaymentExpiredError({
                        expires: credential.challenge.expires,
                    }),
                });
                return { challenge: response, status: 402 };
            }
            // Validate payload structure against method schema
            try {
                method.schema.credential.payload.parse(credential.payload);
            }
            catch (e) {
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error: new Errors.InvalidPayloadError({ reason: e.message }),
                });
                return { challenge: response, status: 402 };
            }
            // User-provided verification (e.g., check signature, submit tx, verify payment).
            // If verification fails, re-issue the challenge so the client can retry.
            let receiptData;
            try {
                receiptData = await verify({ credential, request });
            }
            catch (e) {
                const error = e instanceof Errors.PaymentError
                    ? e
                    : new Errors.VerificationFailedError({ reason: e.message });
                const response = await transport.respondChallenge({
                    challenge,
                    input,
                    error,
                });
                return { challenge: response, status: 402 };
            }
            // If the method's `respond` hook returns a Response, it means this
            // request is a management action (e.g. channel open, voucher POST)
            // and the user's route handler should NOT run. `withReceipt()` will
            // return the management response directly. If undefined, `withReceipt()`
            // expects the caller to pass the user handler's response instead.
            const managementResponse = respond
                ? await respond({ credential, input, receipt: receiptData, request })
                : undefined;
            return {
                status: 200,
                withReceipt(response) {
                    if (managementResponse) {
                        return transport.respondReceipt({
                            receipt: receiptData,
                            response: managementResponse,
                            challengeId: credential.challenge.id,
                        });
                    }
                    if (!response)
                        throw new Error('withReceipt() requires a response argument');
                    return transport.respondReceipt({
                        receipt: receiptData,
                        response: response,
                        challengeId: credential.challenge.id,
                    });
                },
            };
        }, {
            _internal: {
                ...method,
                ...defaults,
                ...options,
                name: method.name,
                intent: method.intent,
                _canonicalRequest: PaymentRequest.fromMethod(method, merged),
            },
        });
    };
}
/**
 * Combines multiple configured payment handlers into a single route handler
 * that presents all methods to the client via multiple `WWW-Authenticate` headers.
 *
 * When no credential is present, all handlers are called and their challenges
 * are merged into a single 402 response. When a credential is present, it is
 * dispatched to the handler matching the credential's `method`+`intent`.
 *
 * @example
 * ```ts
 * import { Mppx, tempo, stripe } from 'mppx/server'
 *
 * const mppx = Mppx.create({
 *   methods: [tempo(), stripe()],
 *   secretKey: process.env.PAYMENT_SECRET_KEY,
 * })
 *
 * app.get('/api/resource', async (req) => {
 *   const result = await Mppx.compose(
 *     mppx['tempo/charge']({ amount: '100', currency: USDC, recipient: '0x...' }),
 *     mppx['stripe/charge']({ amount: '100', currency: 'usd' }),
 *   )(req)
 *   if (result.status === 402) return result.challenge
 *   return result.withReceipt(new Response('OK'))
 * })
 * ```
 */
export function compose(...handlers) {
    if (handlers.length === 0)
        throw new Error('compose() requires at least one handler');
    return async (input) => {
        // Try to extract a Payment credential to decide whether to dispatch or challenge.
        // Only gate on the Payment scheme — other auth schemes (Bearer, Basic, etc.)
        // should fall through to the merged-402 path so all offers are presented.
        const header = input.headers.get('Authorization');
        const paymentHeader = header ? Credential.extractPaymentScheme(header) : null;
        if (paymentHeader) {
            // Parse the credential to find method+intent for dispatch.
            let credential;
            try {
                credential = Credential.deserialize(paymentHeader);
            }
            catch { }
            if (credential) {
                const { method: credMethod, intent: credIntent } = credential.challenge;
                const credReq = credential.challenge.request;
                const credDetails = (credReq.methodDetails ?? {});
                // Filter by name+intent, then narrow by comparing stable request fields
                // from the echoed challenge against each handler's canonical request.
                // Uses the schema-parsed canonical form (not raw options) so that
                // transformed fields (e.g. amount with decimals) match correctly.
                // Also checks inside methodDetails for fields moved there by transforms.
                const candidates = handlers.filter((h) => {
                    const meta = h._internal;
                    if (!meta || meta.name !== credMethod || meta.intent !== credIntent)
                        return false;
                    const canonical = meta._canonicalRequest;
                    if (!canonical)
                        return true;
                    const canonicalDetails = (canonical.methodDetails ?? {});
                    for (const field of ['amount', 'currency', 'recipient', 'chainId']) {
                        const canonicalVal = canonical[field] ?? canonicalDetails[field];
                        if (canonicalVal !== undefined &&
                            String(canonicalVal) !== String(credReq[field] ?? credDetails[field]))
                            return false;
                    }
                    return true;
                });
                const match = candidates[0] ??
                    handlers.find((h) => {
                        const meta = h._internal;
                        return meta?.name === credMethod && meta?.intent === credIntent;
                    });
                if (match)
                    return match(input);
            }
            // Payment credential present but no matching handler — dispatch to first
            // handler which will reject with an appropriate error (invalid challenge, etc.).
            return handlers[0](input);
        }
        // No credential — call all handlers and merge 402 challenges.
        const results = await Promise.all(handlers.map((h) => h(input)));
        // Merge WWW-Authenticate headers from all 402 responses.
        const mergedHeaders = new Headers();
        mergedHeaders.set('Cache-Control', 'no-store');
        let body = null;
        for (const result of results) {
            if (result.status !== 402)
                continue;
            const response = result.challenge;
            const wwwAuth = response.headers.get('WWW-Authenticate');
            if (wwwAuth)
                mergedHeaders.append('WWW-Authenticate', wwwAuth);
            // Use the first handler's body for the problem details response.
            if (!body) {
                const contentType = response.headers.get('Content-Type');
                if (contentType)
                    mergedHeaders.set('Content-Type', contentType);
                body = await response.text();
            }
        }
        return {
            status: 402,
            challenge: new Response(body, { status: 402, headers: mergedHeaders }),
        };
    };
}
/**
 * Wraps a payment handler to create a Node.js HTTP listener.
 *
 * On 402: writes the challenge response and ends the connection.
 * On 200: sets the Payment-Receipt header; caller should write response body.
 *
 * @example
 * ```ts
 * import * as http from 'node:http'
 * import { Mppx } from 'mppx/server'
 *
 * const payment = Mppx.create({ ... })
 *
 * http.createServer(async (req, res) => {
 *   const result = await Mppx.toNodeListener(
 *     payment.charge({
 *       amount: '1', currency: '...', recipient: '0x...',
 *     }),
 *   )(req, res)
 *   if (result.status === 402) return
 *   res.end('OK')
 * })
 * ```
 */
export function toNodeListener(handler) {
    return async (req, res) => {
        const result = await handler(Request.fromNodeListener(req, res));
        if (result.status === 402) {
            await NodeListener.sendResponse(res, result.challenge);
        }
        else {
            const wrapped = result.withReceipt(new globalThis.Response());
            res.setHeader('Payment-Receipt', wrapped.headers.get('Payment-Receipt'));
        }
        return result;
    };
}
