// mppx/dist/server/Transport.js
import * as Challenge from '../Challenge.js';
import * as Credential from '../Credential.js';
import * as Errors from '../Errors.js';
import * as core_Mcp from '../Mcp.js';
import * as Receipt from '../Receipt.js';
export { mcpSdk } from '../mcp-sdk/server/Transport.js';
/**
 * Creates a custom server-side transport.
 *
 * @example
 * ```ts
 * import { Transport } from 'mppx/server'
 *
 * const custom = Transport.from({
 *   name: 'custom',
 *   getCredential(input) { ... },
 *   respondChallenge({ challenge, input }) { ... },
 *   respondReceipt({ receipt, response, challengeId }) { ... },
 * })
 * ```
 */
export function from(transport) {
    return transport;
}
/**
 * HTTP transport for server-side payment handling.
 *
 * - Reads credentials from the `Authorization` header
 * - Issues challenges via `WWW-Authenticate` header with 402 status
 * - Attaches receipts via `Payment-Receipt` header
 */
export function http() {
    return from({
        name: 'http',
        getCredential(request) {
            const header = request.headers.get('Authorization');
            if (!header)
                return null;
            const payment = Credential.extractPaymentScheme(header);
            if (!payment)
                return null;
            return Credential.deserialize(payment);
        },
        respondChallenge({ challenge, error }) {
            const headers = {
                'WWW-Authenticate': Challenge.serialize(challenge),
                'Cache-Control': 'no-store',
            };
            let body = null;
            if (error) {
                headers['Content-Type'] = 'application/problem+json';
                body = JSON.stringify(error.toProblemDetails(challenge.id));
            }
            return new Response(body, { status: error?.status ?? 402, headers });
        },
        respondReceipt({ receipt, response }) {
            const headers = new Headers(response.headers);
            headers.set('Payment-Receipt', Receipt.serialize(receipt));
            return new Response(response.body, {
                status: response.status,
                statusText: response.statusText,
                headers,
            });
        },
    });
}
/**
 * MCP transport for server-side payment handling with raw JSON-RPC.
 *
 * - Reads credentials from `_meta["org.paymentauth/credential"]`
 * - Issues challenges via JSON-RPC error with code -32042/-32043
 * - Attaches receipts via `_meta["org.paymentauth/receipt"]`
 *
 * Use this transport when handling raw JSON-RPC messages directly.
 * For use with `@modelcontextprotocol/sdk`, use `mcpSdk()` instead.
 */
export function mcp() {
    return from({
        name: 'mcp',
        getCredential(request) {
            const meta = request.params?._meta;
            const credential = meta?.[core_Mcp.credentialMetaKey];
            if (!credential)
                return null;
            return credential;
        },
        respondChallenge({ challenge, input, error }) {
            return {
                jsonrpc: '2.0',
                id: input.id,
                error: {
                    code: mcpErrorCode(error),
                    message: error?.message ?? 'Payment Required',
                    data: {
                        httpStatus: error?.status ?? 402,
                        challenges: [challenge],
                        ...(error && { problem: error.toProblemDetails(challenge.id) }),
                    },
                },
            };
        },
        respondReceipt({ receipt, response, challengeId }) {
            if ('error' in response)
                return response;
            const mcpReceipt = {
                ...receipt,
                challengeId,
            };
            return {
                ...response,
                result: {
                    ...response.result,
                    _meta: {
                        ...response.result._meta,
                        [core_Mcp.receiptMetaKey]: mcpReceipt,
                    },
                },
            };
        },
    });
}
/** @internal */
function mcpErrorCode(error) {
    if (!error)
        return core_Mcp.paymentRequiredCode;
    if (error instanceof Errors.MalformedCredentialError)
        return -32602;
    if (error instanceof Errors.PaymentRequiredError)
        return core_Mcp.paymentRequiredCode;
    return core_Mcp.paymentVerificationFailedCode;
}
